src/audio/volume/volume_hifi3.c

```diff
@@ -205,6 +460,127 @@ static void vol_s16_to_s16(struct comp_dev *dev, struct audio_stream *sink,
 				    out, sizeof(ae_int16));
 		}
 	}
+#elif (OPT == OPT1)
+	struct vol_data *cd = comp_get_drvdata(dev);
+	ae_f64 mult;
+	ae_f32x2 volume;
+	ae_f32x2 out_sample;
+	ae_f16x4 in_sample = AE_ZERO16();
+	size_t channel;
+	int i;
+	const int inc = sizeof(ae_int16) * sink->channels;
+
+	/* Processing per channel */
+	for (channel = 0; channel < sink->channels; channel++) {
+
+		ae_int16 *in = (ae_int16 *)source->r_ptr + channel;
+		ae_int16 *out = (ae_int16 *)sink->w_ptr + channel;
+		/* Load volume */
+		volume = (ae_f32x2)cd->volume[channel];
+
+		for (i = 0; i < frames; i++) {
+
+			/* Set source as circular buffer */
+			vol_setup_circular(source);
```

Comment:
this function appears to be duplicated all over the code base for HiFi. @lgirdwood do you know why that is rather than having it in a common spot?

Reply:
@lgirdwood looks like this got closed pre-maturely

Reply:
Good point, @singalsu any idea why we have vol_setup_circular() in multiple places. I'm assuming its a wrapper to setup xtensa HIFI looping ?

Reply:
That function sets the xtensa circular start and end registers. Since there's only single circular modifier in HW the limits need to be every time when accessing source, sink, or the new gain values (no function used but write regs directly) circular buffers.

Reply:
ok, so it seems we need to have a
```c
#define for_each_frame(i, frames, source) \
    for (i = 0; i < frames; i++) { \
			/* Set source as circular buffer */ \
			xtensa_loop_setup(source);
``` 
this would be defined as above for xtensa but would miss out the loop setup for non xtensa.

Reply:
oh, and the above will be a separate PR. @cujomalainey sound Ok for you ? cleans up the looping problem and could be used by all processing components.

Reply:
Yea just wanted to understand the underlying cause :) I am ok with whatever solution is best